polkadot/node/network/collator-protocol/src/validator_side/collation.rs

```diff
@@ -216,27 +221,97 @@ impl CollationStatus {
 }
 
 /// Information about collations per relay parent.
-#[derive(Default)]
 pub struct Collations {
 	/// What is the current status in regards to a collation for this relay parent?
 	pub status: CollationStatus,
 	/// Collator we're fetching from, optionally which candidate was requested.
 	///
-	/// This is the currently last started fetch, which did not exceed `MAX_UNSHARED_DOWNLOAD_TIME`
-	/// yet.
+	/// This is the last fetch for the relay parent. The value is used in
+	/// `get_next_collation_to_fetch` (called from `dequeue_next_collation_and_fetch`) to determine
+	/// if the last fetched collation is the same as the one which just finished. If yes - another
+	/// collation should be fetched. If not - another fetch was already initiated and
+	/// `get_next_collation_to_fetch` will do nothing.
+	///
+	/// For the reasons above this value is not set to `None` when the fetch is done! Don't use it
+	/// to check if there is a pending fetch.
 	pub fetching_from: Option<(CollatorId, Option<CandidateHash>)>,
-	/// Collation that were advertised to us, but we did not yet fetch.
-	pub waiting_queue: VecDeque<(PendingCollation, CollatorId)>,
+	/// Collation that were advertised to us, but we did not yet fetch. Grouped by `ParaId`.
+	waiting_queue: BTreeMap<ParaId, VecDeque<(PendingCollation, CollatorId)>>,
 	/// How many collations have been seconded.
 	pub seconded_count: usize,
+	/// What collations were fetched so far for this relay parent.
+	fetched_per_para: BTreeMap<ParaId, usize>,
+	// Claims per `ParaId` for the assigned core at the relay parent. This information is obtained
+	// from `GroupAssignments` which contains either the claim queue (if runtime supports it) for
+	// the core or the `ParaId` of the parachain assigned to the core.
+	claims_per_para: BTreeMap<ParaId, usize>,
+	// Represents the claim queue at the relay parent. The `bool` field indicates if a candidate
+	// was fetched for the `ParaId` at the position in question. In other words - if the claim is
+	// 'satisfied'. If the claim queue is not available `claim_queue_state` will be `None`.
+	claim_queue_state: Option<Vec<(bool, ParaId)>>,
```

Comment:
> Is that how the proposed system works or have I missed something? Thinking how big of an issue it is but it seems concerning. We might have to clean-up the cq_state in such cases (maybe).

I'm afraid that's correct. We run some checks before enqueuing the candidate but we don't validate it of course. I suppose it shouldn't be hard to craft a garbage candidate which will pass the second checks, claim the slot and prevent legitimate validators from pushing candidates.

> Overall I think it's a good idea to optimistically track what claims should be 'satisfied' already.

It will make more sense to have the 'satisfaction' check based on what's seconded and keep the fetches as 'pending seconding' (similar to the way I handle pending fetches in this PR). We can track this with `Seconded` and `Invalid` messages and accept (and fetch) advertisements until we have enough seconded to 'satisfy' the claim queue. 

A misbehaving collator will still be able to spam the validators but on the first invalid collation it is supposed to get reported (and disconnected maybe? I don't know how this part of the code works).

Reply:
I've reworked the PR to track the number of seconded candidates per para. Additionally I realized that pending items per relay parent can't be more than one which simplifies the PR a little bit.